### PR TITLE
blas/gonum: add Zaxpy

### PR DESCRIPTION
**Pull request by vladimir-ch**
Starting to slowly add complex BLAS routines.

### CODE REVIEW

File: blas/testblas/zaxpy.go [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+			alpha:   3 + 7i,
+			x:       []complex128{2 + 3i, 3 + 4i, 4 + 5i},
+			y:       []complex128{20 + 30i, 30 + 40i, 40 + 50i},
+			want:    []complex128{5 + 53i, 11 + 73i, 17 + 93i},
+			wantRev: []complex128{-3 + 73i, 11 + 73i, 25 + 73i},
+		},
+	} {
+		n := len(test.x)
+		for _, incX := range []int{-10, -1, 1, 10} {
+			for _, incY := range []int{-10, -1, 1, 10} {
```
Comment (btracey) [inline blas/testblas/zaxpy.go]: It's probably a good idea to have these be non-equal between x and y, just in case.

File: blas/testblas/zaxpy.go [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+					for i, v := range test.wantRev {
+						want[i*aincY] = v
+					}
+				}
+
+				impl.Zaxpy(n, test.alpha, x, incX, y, incY)
+
+				prefix := fmt.Sprintf("Case %v (incX=%v,incY=%v):", tc, incX, incY)
+
+				equal := true
```
Comment (btracey) [inline blas/testblas/zaxpy.go]: We're going to have this kind of equality check a lot. Should we start a `complexes` package to correspond to `floats`?

File: blas/testblas/zaxpy.go [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+
+func ZaxpyTest(t *testing.T, impl Zaxpyer) {
+	rnd := rand.New(rand.NewSource(1))
+
+	for tc, test := range []struct {
+		alpha complex128
+		x, y  []complex128
+
+		want    []complex128
+		wantRev []complex128
```
Comment (btracey) [inline blas/testblas/zaxpy.go]: Could you comment this field?

File: blas/testblas/zaxpy.go [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+					aincX = -aincX
+				}
+				aincY := incY
+				if aincY < 0 {
+					aincY = -aincY
+				}
+
+				x := make([]complex128, (n-1)*aincX+1)
+				for i := range x {
+					x[i] = complex(rnd.NormFloat64(), rnd.NormFloat64())
```
Comment (btracey) [inline blas/testblas/zaxpy.go]: Populate with NaN instead?

**Comment by btracey** [comment]
It looks like this game is starting again

File: blas/testblas/zaxpy.go [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+			want:    []complex128{22 + 33i, 33 + 44i, 44 + 55i},
+			wantRev: []complex128{24 + 35i, 33 + 44i, 42 + 53i},
+		},
+		{
+			alpha:   3 + 7i,
+			x:       []complex128{2 + 3i, 3 + 4i, 4 + 5i},
+			y:       []complex128{20 + 30i, 30 + 40i, 40 + 50i},
+			want:    []complex128{5 + 53i, 11 + 73i, 17 + 93i},
+			wantRev: []complex128{-3 + 73i, 11 + 73i, 25 + 73i},
+		},
```
Comment (Kunde21) [inline blas/testblas/zaxpy.go]: The main loop in `asm/c128` is 4-wide and the future AVX version will be 8-wide, so there should be test entries of length 10+ in here.

File: blas/testblas/zaxpy.go (hunk 2 above)
Comment (Kunde21) [inline blas/testblas/zaxpy.go]: We had discussed this in the [major breaking changes thread](https://groups.google.com/forum/#!topic/gonum-dev/4oLmSvtDuZY) but didn't come to an agreement on a single package (`Vector/<type>`?) or mirrored packages (`floats` and `complexes`).  If we're going to start building out the complex blas routines, we should get the vector routines organized, too.

File: blas/testblas/zaxpy.go (hunk 3 above)
Comment (vladimir-ch) [inline blas/testblas/zaxpy.go]: Done.

File: blas/testblas/zaxpy.go (hunk 1 above)
Comment (vladimir-ch) [inline blas/testblas/zaxpy.go]: Done.

File: blas/testblas/zaxpy.go (hunk 4 above)
Comment (vladimir-ch) [inline blas/testblas/zaxpy.go]: Done.

File: blas/testblas/zaxpy.go (hunk 2 above)
Comment (vladimir-ch) [inline blas/testblas/zaxpy.go]: At the moment I will do it manually. If we pre-fill the slices with NaNs, `complexes.Equal` would not work here anyway.

**Comment by vladimir-ch** [comment]
PTAL

File: blas/testblas/zaxpy.go (hunk 2 above)
Comment (btracey) [inline blas/testblas/zaxpy.go]: Could you at least make it a function in here?

Comment (btracey) [inline blas/testblas/zaxpy.go]: And `complexes.Same` presumably would work.

Comment (vladimir-ch) [inline blas/testblas/zaxpy.go]: Sorry, I don't know our own packages, that would of course work.
Done.

File: blas/testblas/zaxpy.go [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+			prefix := fmt.Sprintf("Case %v (incX=%v,incY=%v):", tc, incX, incY)
+
+			if !zsame(x, xCopy) {
+				t.Errorf("%v: unexpected modification of x", prefix)
+			}
+
+			if !zsame(y, want) {
+				t.Errorf("%v: unexpected y:\nwant %v\ngot %v", prefix, want, y)
+			}
+			t.Log(prefix)
```
Comment (kortschak) [inline blas/testblas/zaxpy.go]: What is this for?

Comment (vladimir-ch) [inline blas/testblas/zaxpy.go]: Sorry, removed now.

**Pull request merged**